src/zircon/metadatacache/metadata.go

```diff
@@ -0,0 +1,351 @@
+func entryNumberToOffset(entryN uint) uint32 {
+	return uint32(entryN)*apis.EntrySize + apis.BitsetSize
+}
+
+func entryAndBlockToChunkNum(metachunk apis.MetadataID, index uint) apis.ChunkNum {
+	if metachunk == 0 || index >= (1 << apis.EntriesPerBlock) {
+		panic("broken invariant for chunk location")
+	}
+	return apis.ChunkNum(uint64(metachunk << apis.EntriesPerBlock) | uint64(index))
+}
+
+// Allocate a new metadata entry and corresponding chunk number
+func (mc *metadatacache) NewEntry() (apis.ChunkNum, error) {
+	for {
+		metachunk, index, err := mc.findAnyFreeChunk()
+		if err != nil {
+			return 0, err
+		}
+
+		noclobber, err := mc.updateBitset(metachunk, index, true)
+		if err != nil {
+			return 0, err
+		}
+
+		if noclobber {
+			return entryAndBlockToChunkNum(metachunk, index), nil
+		}
+		// welp... guess we gotta go around again! someone else messed with our chunk.
+	}
+}
+
+// Checks whether a chunk has been allocated or not in the bitset part of a certain metachunk.
+func (mc *metadatacache) getBitset(metachunk apis.MetadataID, index uint) (bool, error) {
+	data, _, _, err := mc.leasing.Read(metachunk)
+	if err != nil {
+		return false, err
+	}
+
+	return getBitsetInData(data[0:apis.BitsetSize], index), nil
+}
+
+// Checks whether a chunk has been allocated or not in a bitset.
+func getBitsetInData(bitset []byte, index uint) (bool) {
+	cell := bitset[index / 8]
+	var mask byte = 1 << (index % 8)
+	return (cell & mask) != 0
+}
+
+// Provides write parameters to update a bitset: (offset, data)
+func updateBitsetInData(bitset []byte, index uint, value bool) (uint32, []byte) {
+	cell := bitset[index / 8]
+	var mask byte = 1 << (index % 8)
+	if value {
+		cell |= mask
+	} else {
+		cell &= mask ^ 0xFF
+	}
+	return uint32(index / 8), []byte{ cell }
+}
+
+// Update whether a chunk has been allocated or not in the bitset part of a certain metachunk.
+// Returns 'true' if the request succeeded, false if it was clobbered, and error if anything else happened.
+func (mc *metadatacache) updateBitset(metachunk apis.MetadataID, index uint, value bool) (bool, error) {
+	data, version, _, err := mc.leasing.Read(metachunk)
+	if err != nil {
+		return false, err
+	}
+
+	bitset := data[0:apis.BitsetSize]
+
+	existingValue := getBitsetInData(data[0:apis.BitsetSize], index)
+	if existingValue == value {
+		return false, nil
+	}
+
+	offset, newData := updateBitsetInData(bitset, index, value)
+
+	_, _, err = mc.leasing.Write(metachunk, version, offset, newData)
+	if err != nil {
+		return false, err
+	}
+
+	return true, nil
+}
+
+// Tries to find a free chunk anywhere. Returns (metadataID, index, error)
+func (mc *metadatacache) findAnyFreeChunk() (apis.MetadataID, uint, error) {
+	// First, see if there is an open spot in a lease that we hold
+	metadataID, index, found, err := mc.findAnyLeasedFreeChunk()
+	if err != nil {
+		return 0, 0, err
+	}
+	if found {
+		return metadataID, index, nil
+	}
+
+	// Now start trying everything else
```

Comment:
If there isn't a free chunk, we should potentially release the lease, as I can envision a scenario where on restart one frontend acquires all of the leases, if many are completely full

Reply:
I'll add this one as a TODO.